src/Algebra/Properties/Semigroup/Reasoning.agda

```diff
@@ -0,0 +1,110 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Equational reasoning for semigroups
+-- (Utilities for associativity reasoning, pulling and pushing operations)
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+open import Algebra.Bundles using (Semigroup)
+
+module Algebra.Properties.Semigroup.Reasoning {o ℓ} (S : Semigroup o ℓ) where
+
+open import Algebra.Structures using (IsMagma)
+
+open Semigroup S
+  using (Carrier; _∙_; _≈_; setoid; trans; refl; sym; assoc; ∙-cong; isMagma
+  ; ∙-congˡ; ∙-congʳ)
+open import Relation.Binary.Reasoning.Setoid setoid
+
+private
+ variable
+    u v w x y z : Carrier
+
+
+module Assoc4  {u v w x : Carrier} where
+  [uv∙w]x≈u[vw∙x] : ((u ∙ v) ∙ w) ∙ x ≈ u ∙ ((v ∙ w) ∙ x)
+  [uv∙w]x≈u[vw∙x] = trans (∙-congʳ (assoc u v w)) (assoc u (v ∙ w) x)
+
+  [uv∙w]x≈u[v∙wx] : ((u ∙ v) ∙ w) ∙ x ≈ u ∙ (v ∙ (w ∙ x))
+  [uv∙w]x≈u[v∙wx] = trans (assoc (u ∙ v) w x) (assoc u v (w ∙ x))
+
+  [u∙vw]x≈uv∙wx : (u ∙ (v ∙ w)) ∙ x ≈ (u ∙ v) ∙ (w ∙ x)
+  [u∙vw]x≈uv∙wx = trans (sym (∙-congʳ (assoc u v w))) (assoc (u ∙ v) w x)
+
+  [u∙vw]x≈u[v∙wx] : (u ∙ (v ∙ w)) ∙ x ≈ u ∙ (v ∙ (w ∙ x))
+  [u∙vw]x≈u[v∙wx] = trans (assoc u (v ∙ w) x) (∙-congˡ (assoc v w x))
+
+  uv∙wx≈[u∙vw]x : (u ∙ v) ∙ (w ∙ x) ≈ (u ∙ (v ∙ w)) ∙ x
+  uv∙wx≈[u∙vw]x = trans (sym (assoc (u ∙ v) w x)) (∙-congʳ (assoc u v w))
+
+  uv∙wx≈u[vw∙x] : (u ∙ v) ∙ (w ∙ x) ≈ u ∙ ((v ∙ w) ∙ x)
+  uv∙wx≈u[vw∙x] = trans (assoc u v (w ∙ x)) (∙-congˡ (sym (assoc v w x)))
+
+  u[vw∙x]≈[uv∙w]x : u ∙ ((v ∙ w) ∙ x) ≈ ((u ∙ v) ∙ w) ∙ x
+  u[vw∙x]≈[uv∙w]x = sym (trans (∙-congʳ (assoc u v w)) (assoc u (v ∙ w) x))
+
+  u[vw∙x]≈uv∙wx : u ∙ ((v ∙ w) ∙ x) ≈ (u ∙ v) ∙ (w ∙ x)
+  u[vw∙x]≈uv∙wx = trans (∙-congˡ (assoc v w x)) (sym (assoc u v (w ∙ x)))
+
+  u[v∙wx]≈[uv∙w]x : u ∙ (v ∙ (w ∙ x)) ≈ ((u ∙ v) ∙ w) ∙ x
+  u[v∙wx]≈[uv∙w]x = sym (trans (assoc (u ∙ v) w x) (assoc u v (w ∙ x)))
+
+  u[v∙wx]≈[u∙vw]x : u ∙ (v ∙ (w ∙ x)) ≈ (u ∙ (v ∙ w)) ∙ x
+  u[v∙wx]≈[u∙vw]x  = sym (trans (assoc u (v ∙ w) x) (∙-congˡ (assoc v w x)))
+
+open Assoc4 public
+
+module Pulls (uv≡w : u ∙ v ≈ w) where
```

Comment:
Suggest you don't use `uv≡w` as the name of a proof of `u ∙ v ≈ w`... but use `uv≈w` instead?
(`≡` should only ever be reserved for `PropositionalEquality`)